### PR TITLE
Allow PDep reactions and non-template library reactions in PDep networks

### PR DESCRIPTION
**Pull request by alongd**
This PR allows PDep networks to use kinetics from a LibraryReaction in two cases:
1. When a unimolecular library reaction does not correspond to any kinetics family (but is flagged as `elementary`)
2. When a unimolecular library reaction has PDep kinetics, in which case a high pressure limit rate is generated and saved as a new attribute of the Reaction. This rate is then used in the network.

A test was added for the high pressure limit kinetic generation. This PR involved relatively modest changed to pdep.py, so direct pdep tests were currently not added.

### CODE REVIEW

**Comment by alongd** [comment]
Addresses #1313

**Comment by alongd** [comment]
Tests overall look good on this branch:
We have the usual non deterministic GAV interchange of [`ring(Cyclopropene)` and `ring(oxirene)`] and [`ring(cyclopropanone)` and `ring(2(co)oxirane)`].
The MCH test generated identical models, but then says `Observable species CC1CCCCC1 varied by more than 0.050 on average between old model MCH(1) and new model MCH(1) in condition 1.` The reason is unclear to me.

[main_log.txt](https://github.com/ReactionMechanismGenerator/RMG-Py/files/1987827/main_log.txt)

File: rmgpy/data/kinetics/libraryTest.py [hunk 1]
```diff
@@ -81,4 +83,31 @@ def testSaveLibrary(self):
...
+
+    def test_generate_high_p_limit_kinetics(self):
+        """
+        Test that a :class:Arrhenius kinetics object representing the high pressure limit rate
+        is returned from Troe/Lindmann/PDepArrhenius/Chebyshev kinetic classes
+        """
+        libRxns = self.libraries['lib_net'].getLibraryReactions()
+        for rxn in libRxns:
+            self.assertIsNone(rxn.network_kinetics)
+            logging.debug(rxn)
```
Comment (mjohnson541) [inline rmgpy/data/kinetics/libraryTest.py]: If this debug statement was intended to be included I think you should make the message more specific

File: rmgpy/data/kinetics/libraryTest.py [hunk 2]
```diff
@@ -81,4 +83,31 @@ def testSaveLibrary(self):
...
+                # or if it is a PDepArrhenius or Chebyshev with Pmax < 90 bar
+                self.assertFalse(success)
+            else:
+                self.assertTrue(success)
+                if isinstance(rxn.kinetics, Arrhenius):
+                    # If the library reaction is already an Arrhenius expression, network_kinetics isn't generated
+                    self.assertIsNone(rxn.network_kinetics)
+                else:
+                    self.assertTrue(isinstance(rxn.network_kinetics, Arrhenius))
+            if isinstance(rxn.kinetics,Troe):
```
Comment (mjohnson541) [inline rmgpy/data/kinetics/libraryTest.py]: Should this be under the else immediately before it?

File: rmgpy/data/kinetics/libraryTest.py [hunk 3]
```diff
@@ -81,4 +83,31 @@ def testSaveLibrary(self):
...
+                # some simple quantitative checks for network_kinetics
+                self.assertAlmostEqual(rxn.network_kinetics.n.value_si, 0)
+                self.assertAlmostEqual(int(rxn.network_kinetics.A.value_si), 6e+8)
```
Comment (mjohnson541) [inline rmgpy/data/kinetics/libraryTest.py]: Can you explain these checks with comments or at least in the commit?

File: rmgpy/test_data/testing_database/kinetics/libraries/lib_net/reactions.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+entry(
+    index = 2,
+    label = "H + CH3 <=> CH4",
+    degeneracy = 6.0,
+    kinetics = Arrhenius(
+        A = (2.94005e-11, 'm^3/(mol*s)'),
+        n = 5.135,
+        Ea = (33.0118, 'kJ/mol'),
+        T0 = (1, 'K')),
+    elementary_high_p = True,
```
Comment (mjohnson541) [inline rmgpy/test_data/testing_database/kinetics/libraries/lib_net/reactions.py]: Why isn't it necessary to label most of the other reactions this way?

File: rmgpy/rmg/model.py [hunk 5]
```diff
@@ -1429,7 +1429,16 @@ def addSeedMechanismToCore(self, seedMechanism, react=False):
                  reversible=rxn.reversible
                  )
             r, isNew = self.makeNewReaction(rxn) # updates self.newSpeciesList and self.newReactionlist
-            if not isNew: logging.info("This library reaction was not new: {0}".format(rxn))
+            if not isNew:
```
Comment (mjohnson541) [inline rmgpy/rmg/model.py]: Why do we need this processing changes for seeds and reaction libraries?

File: documentation/source/users/rmg/database/modification.rst [hunk 6]
```diff
@@ -81,6 +81,25 @@ To conform to RMG's format, simply copy and modify an existing library.
...
+(i.e., not the measured rate at some low or medium experimental pressure). If set to ``True``, RMG will use the high pressure
+limit rate when constructing pressure-dependent networks. The kinetics entry of the original library reaction will only be
+updated if it is an Arrhenius type (will be replaced with either PDepArrhenius or Chebyshev, as specified in the
+pressureDependent block of the input file). If set to ``False`` (the default value), RMG will not use the high pressure
+limit rate in network exploration, and will not convert Arrhenius kinetics of library reactions that have no template
+(a corresponding reaction family) into a pressure-dependent form.
+
+2. **has_pdep_route**: If set to ``True`` and RMG discovers a pressure-dependent reaction with the same reactants and products,
+the latter *will* be considered in addition to the library reaction. This is useful for cases when more than one pathway connects
+the samethe same reactants and products, and some of these pathways are well-skipping reactions. If set to ``False``,
```
Comment (mjohnson541) [inline documentation/source/users/rmg/database/modification.rst]: typo

File: documentation/source/users/rmg/database/modification.rst [hunk 7]
```diff
@@ -81,6 +81,25 @@ To conform to RMG's format, simply copy and modify an existing library.
...
+similar network reactions will not be considered in the model generation.
```
Comment (mjohnson541) [inline documentation/source/users/rmg/database/modification.rst]: What is the default value?

File: documentation/source/users/rmg/database/modification.rst [hunk 8]
```diff
@@ -81,6 +81,25 @@ To conform to RMG's format, simply copy and modify an existing library.
...
+and/or product) with a kinetics entry that has information about the high pressure kinetics, i.e., Troe or Lindemann,
+PDepArrhenius or Chebyshev that are defined up to at least 100 bar, or Arrhenius that represents the high pressure limit
+(i.e., not the measured rate at some low or medium experimental pressure). If set to ``True``, RMG will use the high pressure
+limit rate when constructing pressure-dependent networks. The kinetics entry of the original library reaction will only be
+updated if it is an Arrhenius type (will be replaced with either PDepArrhenius or Chebyshev, as specified in the
+pressureDependent block of the input file). If set to ``False`` (the default value), RMG will not use the high pressure
+limit rate in network exploration, and will not convert Arrhenius kinetics of library reactions that have no template
+(a corresponding reaction family) into a pressure-dependent form.
+
+2. **has_pdep_route**: If set to ``True`` and RMG discovers a pressure-dependent reaction with the same reactants and products,
```
Comment (mjohnson541) [inline documentation/source/users/rmg/database/modification.rst]: Would "allow_pdep_route" be a better name?

File: rmgpy/test_data/testing_database/kinetics/libraries/lib_net/reactions.py (hunk 4 above)
Comment (alongd) [reply rmgpy/test_data/testing_database/kinetics/libraries/lib_net/reactions.py]: It is, you are right. I wasn't consistent in defining `generate_high_p_limit_kinetics()`.
I added a check for `elementary_high_p` now.

File: rmgpy/rmg/model.py (hunk 5 above)
Comment (alongd) [reply rmgpy/rmg/model.py]: To convert each new reaction from libraries/seeds which has an Arrhenius-type kinetics (and is flagged as high-P elementary) into a PDep-type kinetics
(`processNewReactions` calls `rxn.generate_high_p_limit_kinetics()`)

**Comment by alongd** [comment]
Thanks, @mjohnson541 
I addressed your comments in fixup commits. The last commit isn't a fixup, since `allow_pdep_route` isn't part of this PR (I just added the documentation for it here).
Let me know when to squash

**Comment by mjohnson541** [comment]
Ok this looks good...even Codacy is happy!

**Comment by alongd** [comment]
Sweet! OK, rebased and squashed, ready to merge once tests pass (please re-approve)

**Comment by mliu49** [comment]
I think you'll need to rebase and reapprove again since you just merged the other PR...

**Comment by mjohnson541** [comment]
Yes, but that doesn't seem to have prevented me from approving it again...

**Comment by alongd** [comment]
Please re-approve... :)

**Pull request merged**